de.variantsync.core/src/main/java/de/variantsync/core/ast/AST.java

```diff
@@ -0,0 +1,113 @@
+package de.variantsync.core.ast;
+
+import java.util.*;
+
+import com.google.gson.annotations.Expose;
+import com.google.gson.annotations.SerializedName;
+import de.variantsync.core.interfaces.Grammar;
+
+public class AST<G extends Grammar, Value> {
+
+	@Expose
+	@SerializedName(value = "uuid")
+	private UUID id;
+	@Expose
+	@SerializedName(value = "value")
+	private Value value;
+	@Expose
+	@SerializedName(value = "grammar_type")
+	private G type;
+	@Expose
+	@SerializedName(value = "children")
+	private List<AST<G, Value>> children;
+
+	private final String INDENT_STRING = "    ";
+
+	@Override
+	public String toString() {
+		StringBuilder result = new StringBuilder();
+		if (value == null) {
+			return result.toString();
+		} else {
+			int[] level = { 0 }; // pointer magic
+
+			HashSet<Integer> levelFinished = new HashSet<>(); // eg. is level 3 finished?
+			toString(result, this, level, levelFinished, false);
+		}
+
+		return result.toString();
+	}
+
+	private void toString(StringBuilder result, AST<G, Value> parent, int[] level, HashSet<Integer> levelFinished, boolean isLast) {
+		for (int i = 0; i < level[0]; i++) {
+			String toAppend = INDENT_STRING + "\u2502 ";
+			if (levelFinished.contains(i)) {
+				// no need of signs like | because of depth
+				toAppend = INDENT_STRING + "  ";
+			}
+			if (i == level[0] - 1) {
+				// end of indent make arrow
+				toAppend = INDENT_STRING + "\u251C\u2500";
+				if (isLast) {
+					// last child of parent, no arrow down needed
+					toAppend = INDENT_STRING + "\u2514\u2500";
+				}
+
+			}
+			result.append(toAppend);
+			if (i != 0) result.append(" ");
+
+		}
+
+		result.append(parent.type).append(" ").append(parent.value).append(" Depth: ").append(level[0]).append("\n");
+		level[0]++;
+		for (AST<G, Value> child : parent.children) {
+			isLast = false;
+			if (parent.children.indexOf(child) == parent.children.size() - 1) {
+				// last child of subtree, meaning level finished, needs |
+				levelFinished.add(level[0] - 1);
+				isLast = true;
+			} else if (parent.children.indexOf(child) == 0) {
+				// first child of new sub tree with unfinished level needs |
+				levelFinished.remove(level[0] - 1);
+			}
+			toString(result, child, level, levelFinished, isLast);
+
+		}
+		level[0]--;
+	}
+
+	public AST(G type, Value value) {
+		this.id = UUID.randomUUID();
+		this.type = type;
+		this.value = value;
+		this.children = new ArrayList<>();
+	}
```

Comment:
Please add a second constructor were the id is passed as argument and make this constructor use the new one.